File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return helper(root->left, LONG_MIN, root->val) && 
                    helper(root->right, root->val, LONG_MAX); 
    }
    
    bool helper(TreeNode *root, long lower, long upper) {
        if(root == nullptr) {
            return true;
        }
        else {

            
            if(root->val < upper && root->val > lower) {
                
                return helper(root->left, lower, root->val) 
                    && helper(root->right, root->val ,upper);
            }
            else {
                return false;
            }
        }
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(curr, minVal, maxVal):
            if curr is None:
                return True
            
            valid_l = False
            valid_r = False
            
            # left branch
            if curr.left is None:
                valid_l = True
            elif curr.left.val > minVal and curr.left.val < maxVal and curr.left.val < curr.val:
                valid_l = dfs(curr.left, minVal, curr.val)                
            else:
                valid_l = False
            
            # right branch
            if curr.right is None:
                valid_r = True
            elif curr.right.val < maxVal and curr.right.val > minVal and curr.right.val > curr.val:
                valid_r = dfs(curr.right, curr.val, maxVal)
            else:
                valid_r = False

            return valid_l and valid_r
            
        return dfs(root, -sys.maxsize, sys.maxsize)
        